chore!: migrate to Node 18

MarkDuckworth opened the pull request:
Fixes b/397525375. Uses steps defined in b/397525375.

MarkDuckworth committed:
Fixes for google-gax major version update

MarkDuckworth committed:
Update to node 18+. Update all deps, except chai and chai-as-promised

MarkDuckworth committed:
formatting

MarkDuckworth committed:
more formatting

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

sofisl reviewed (commented):
Just want to make sure you're following the script here, b/397525375 (might be worth mentioning it in your PR description)

MarkDuckworth commented on .github/.OwlBot.lock.yaml:
```diff
@@ -13,5 +13,4 @@
   image: gcr.io/cloud-devrel-public-resources/owlbot-nodejs:latest
-  digest: sha256:0d39e59663287ae929c1d4ccf8ebf7cef9946826c9b86eda7e85d8d752dbb584
-# created: 2024-11-21T22:39:44.342569463Z
+  digest: sha256:ebf1487fdb5be0d02d49a20b01547be3cd15cbd03f4ded7b47c65eae7920a080
```
Thank you

MarkDuckworth committed:
Update .OwlBot.lock.yaml

Update to latest image

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

MarkDuckworth committed:
Update opentelemetry packages to v2

MarkDuckworth committed:
Merge branch 'node-18' of github.com:googleapis/nodejs-firestore into node-18

MarkDuckworth committed:
fix for opentelemetry dep tree with override

MarkDuckworth committed:
Remove ignore of no-floating-promises lint error

MarkDuckworth commented on .github/sync-repo-settings.yaml:
```diff
@@ -8,9 +8,9 @@ branchProtectionRules:
-      - test (16)
       - test (18)
+      - test (20)
+      - test (22)
```
done

MarkDuckworth committed:
Floating promise lint error fixes (#2400)

* I will fix the floating promise errors.

This PR fixes all the floating promise errors in the test files. It also converts all the `then` calls to `async/await` in the test files.

* remove jules txt output files

* Fix hanging promise errors

* Attempting CI fix of tests

* Change solution to floating promises where adding await previously changed the meaning of the test.

* revert void assert.rejects changes in gapic_firestore_v1beta1.ts tests

* Address PR comments

---------

Co-authored-by: google-labs-jules[bot] <161369871+google-labs-jules[bot]@users.noreply.github.com>

sofisl commented:
Approved! I think your tests are failing because you don't have a Node 18 kokoro file (this might be because it's excluded in the owlbot.py file). It should mostly just comprise of renaming the kokoro 14 files to 18 (and making sure they're using 18 as the image, node version, etc.).

MarkDuckworth committed:
Merge branch 'main' into node-18

MarkDuckworth committed:
Merge branch 'main' into node-18

MarkDuckworth committed:
linting

MarkDuckworth committed:
Update .OwlBot.lock.yaml

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

MarkDuckworth committed:
update @google-cloud/opentelemetry-cloud-trace-exporter

MarkDuckworth committed:
Skip tests failing because of gax-fallback regression

MarkDuckworth committed:
Merge branch 'main' into node-18

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

MarkDuckworth commented on dev/src/v1/firestore_admin_client.ts:
```diff
@@ -3749,8 +3749,8 @@ export class FirestoreAdminClient {
-    options?: gax.CallOptions
-  ): AsyncIterable<protos.google.longrunning.ListOperationsResponse> {
+    options?: gax.CallOptions,
+  ): AsyncIterable<protos.google.longrunning.IOperation> {
```
The gax client was updated to return IOperation. This was a correction of the return type. See: https://github.com/googleapis/gax-nodejs/commit/c5af911c5d4eb9caee8fbe6865d58ad904d2d626

We have to update our client too.

MarkDuckworth committed:
Get the API report building again

MarkDuckworth commented on api-report/firestore.api.md:
```diff
@@ -1,1714 +0,0 @@
-## API Report File for "@google-cloud/firestore"
```
The script started failing with the upgrade. It was accidentally deleted from git. I've fixed the script and re-added the file

MarkDuckworth committed:
add renamed api-report script
